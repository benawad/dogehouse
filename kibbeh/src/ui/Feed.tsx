import React, { MouseEventHandler, ReactNode } from "react";
import { RoomCardProps, RoomCard } from "./RoomCard";
import { Button } from "./Button";

export interface FeedProps {
  title: string;
  actionTitle: string;
  onActionClicked: MouseEventHandler<HTMLButtonElement>;
  rooms: RoomCardProps[];
  emptyPlaceholder: ReactNode;
}

export const Feed: React.FC<FeedProps> = ({
  title,
  actionTitle,
  onActionClicked,
  rooms,
  emptyPlaceholder,
}) => {
  return (
    <div className="flex-1 flex-col">
      <div className="justify-between items-end mb-5">
        <h4 className="text-primary-100 ml-4">{title}</h4>
        <Button onClick={onActionClicked}>{actionTitle}</Button>
      </div>
      <div className="flex-col space-y-4">
        {rooms.length === 0 && emptyPlaceholder}
        {/* <hr className="border-primary-600" /> */}
        {rooms.map((room, index) => (
          <RoomCard
            key={index}
            title={room.title}
            subtitle={room.subtitle}
            scheduledFor={room.scheduledFor}
            listeners={room.listeners}
            tags={room.tags}
          />
        ))}
      </div>
    </div>
  );
};
