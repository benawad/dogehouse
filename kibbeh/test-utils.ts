import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import "mutationobserver-shim";

const customRender = (ui: any, options = {}) =>
  render(ui, {
    ...options,
  });

export * from "@testing-library/react";

export { customRender as render };
