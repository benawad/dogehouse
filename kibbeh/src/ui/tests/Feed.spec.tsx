// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck @todo this file needs to be fixed

import React from "react";

import { render } from "../../../test-utils";
import { Feed, FeedProps } from "../Feed";

jest.useFakeTimers();
// eslint-disable-next-line init-declarations
let defaultProps: FeedProps;

describe("Feed", () => {
  beforeEach(() => {
    defaultProps = {
      title: "Your feed",
      actionTitle: "New Room",
      emptyPlaceholder: <div>Empty</div>,
      onActionClicked: () => jest.fn(),
      rooms: [],
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const { getByTestId, getByText } = render(<Feed {...defaultProps} />);
    const component = getByTestId("feed");
    const title = getByText(defaultProps.title);
    const actionTitle = getByText(defaultProps.actionTitle);

    expect(component).toBeVisible();
    expect(title).toBeVisible();
    expect(actionTitle).toBeVisible();
  });

  it("should render empty placeholder if rooms length is 0", () => {
    const { getByText } = render(<Feed {...defaultProps} />);
    const placeholder = getByText("Empty");

    expect(placeholder).toBeVisible();
  });

  it("should render one room if room", async () => {
    defaultProps = {
      ...defaultProps,
      rooms: [
        {
          id: 1,
          name: "my room",
        },
      ],
    };
    const { findAllByText } = render(<Feed {...defaultProps} />);
    const rooms = await findAllByText("my room");

    expect(rooms).toHaveLength(1);

    let placeholder = null;
    try {
      placeholder = await findAllByText("Empty");
    } catch (error) {
      expect(placeholder).toBeNull();
    }
  });

  it("should match snapshot", () => {
    const { getByTestId } = render(<Feed {...defaultProps} />);
    const component = getByTestId("feed");

    expect(component).toMatchSnapshot();
  });
});
