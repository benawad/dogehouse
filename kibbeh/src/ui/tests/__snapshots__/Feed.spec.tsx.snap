// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed should match snapshot 1`] = `
<div
  class="flex-1 flex-col"
  data-testid="feed"
>
  <div
    class="justify-between items-end mb-5"
  >
    <h4
      class="text-primary-100 mb-auto"
    >
      Your feed
    </h4>
    <button
      class="py-2 px-6 text-sm rounded-lg text-button bg-accent hover:bg-accent-hover disabled:text-accent-disabled disabled:bg-accent-hover font-bold flex items-center justify-center "
      data-testid="button"
    >
      <span
        class="flex items-center"
      >
        New Room
      </span>
    </button>
  </div>
  <div
    class="flex-col space-y-4"
  >
    <div>
      Empty
    </div>
  </div>
</div>
`;
